src/torchmetrics/functional/clustering/utils.py

```diff
@@ -0,0 +1,77 @@
+# Copyright The Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+import torch
+from torch import Tensor
+
+from torchmetrics.utilities.checks import _check_same_shape
+
+
+def calculate_contingency_matrix(
+    preds: Tensor, target: Tensor, eps: Optional[float] = 1e-16, sparse: bool = False
+) -> Tensor:
+    """Calculate contingency matrix.
+
+    Args:
+        preds: predicted labels
+        target: ground truth labels
+        eps: value added to contingency matrix
+        sparse: If True, returns contingency matrix as a sparse matrix. Else, return as dense matrix.
+            `eps` must be `None` if `sparse` is `True`.
+
+    Returns:
+        contingency: contingency matrix of shape (n_classes_target, n_classes_preds)
+
+    """
+    if eps is not None and sparse is True:
+        raise ValueError("Cannot specify `eps` and return sparse tensor.")
+
+    preds_classes, preds_idx = torch.unique(preds, return_inverse=True)
+    target_classes, target_idx = torch.unique(target, return_inverse=True)
+
+    n_classes_preds = preds_classes.size(0)
+    n_classes_target = target_classes.size(0)
+
+    contingency = torch.sparse_coo_tensor(
+        torch.stack((target_idx, preds_idx)), torch.ones(target_idx.size(0)), (n_classes_target, n_classes_preds)
+    )
+
+    if not sparse:
+        contingency = contingency.to_dense()
+    if eps:
+        contingency = contingency + eps
+
+    return contingency
+
+
+def check_cluster_labels(preds: Tensor, target: Tensor) -> None:
+    """Check shape of input tensors and if they are real, discrete tensors.
+
+    Args:
+        preds: predicted labels
+        target: ground truth labels
+
+    """
+    _check_same_shape(preds, target)
```

Comment:
pretty sure we need to secure this at some point for the implementation to work
```suggestion
    _check_same_shape(preds, target)
    if preds.ndim != 1:
        raise ValueError(f"Expected arguments to be 1d tensors but got {preds.ndim} and {target.ndim}")
```